modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,596 @@
+
+            if (!F.isEmpty(results)) {
+                Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> results0 = mapPartsHashes(results, ctx.req.nodes());
+
+                IdleVerifyResultV2 chkRes = SnapshotChecker.reduceHashesResults(results0, errors0);
+
+                clusterOpFut.onDone(new SnapshotPartitionsVerifyTaskResult(ctx.clusterMetas, chkRes));
+            }
+            else {
+                assert !errors0.isEmpty();
+
+                clusterOpFut.onDone(new IgniteSnapshotVerifyException(errors0));
+            }
+        }
+
+        return new GridFinishedFuture<>();
+    }
+
+    /** Phase 2 beginning.  */
+    private IgniteInternalFuture<SnapshotCheckResponse> validateParts(SnapshotCheckProcessRequest req) {
+        SnapshotCheckContext ctx;
+
+        // The context can be null, if a required node leaves before this phase.
+        if (!req.nodes().contains(kctx.localNodeId()) || (ctx = context(null, req.requestId())) == null || ctx.locMeta == null)
+            return new GridFinishedFuture<>();
+
+        File snpDir = kctx.cache().context().snapshotMgr().snapshotLocalDir(req.snapshotName(), req.snapshotPath());
+
+        ctx.fut = processPartitions(ctx.locMeta, snpDir, req.groups(), req.allRestoreHandlers, true);
+
+        return ctx.fut;
+    }
+
+    /** */
+    private GridFutureAdapter<SnapshotCheckResponse> processPartitions(
+        SnapshotMetadata meta,
+        File snpDir,
+        @Nullable Collection<String> groups,
+        boolean allRestoreHandlers,
+        boolean check
+    ) {
+        GridFutureAdapter<SnapshotCheckResponse> fut = new GridFutureAdapter<>();
+
+        CompletableFuture<? extends Map<?, ?>> workingFut;
+
+        IgniteSnapshotManager snpMgr = kctx.cache().context().snapshotMgr();
+
+        if (allRestoreHandlers) {
+            workingFut = CompletableFuture.supplyAsync(() -> {
+                try {
+                    return snpMgr.handlers().invokeAll(SnapshotHandlerType.RESTORE,
+                        new SnapshotHandlerContext(meta, groups, kctx.cluster().get().localNode(), snpDir, false, check));
+                }
+                catch (IgniteCheckedException e) {
+                    throw new IgniteException("Failed to call custom snapshot validation handlers.", e);
+                }
+            });
+        }
+        else
+            workingFut = snpMgr.checker().checkPartitions(meta, snpDir, groups, false, check, false);
+
+        workingFut.whenComplete((res, err) -> {
+            if (err != null)
+                fut.onDone(err);
+            else
+                fut.onDone(new SnapshotCheckResponse(res));
+        });
+
+        return fut;
+    }
+
+    /**
+     * Call snapshot all the registered validaton handlers.
+     *
+     * @see IgniteSnapshotManager#handlers()
+     * @see #checkCustomHandlersResults(String, Map)
+     */
+    GridFutureAdapter<SnapshotCheckResponse> invokeCustomHandlers(
```

Comment:
This method isn't used in this class. It looks like `SnapshotHandlerRestoreJob` should call method from the `SnapshotChecker` instead